[distributed] Make rref_proxy._invoke_rpc trully async when needed.

kumpera opened the pull request:
Summary:

From #67626: RRefProxy (rref.rpc_async, rref.rpc_sync, rref.remote) currently uses a blocking RPC call to the owner

This is done by chaining async calls. In the sync case we wait on the
resulting Future.

Test Plan:

I ran rpc_tests using tensorpipe_rpc_agent_test_fixture.py and had to
adjust test_rref_proxy_timeout to the new behavior.

I ran into test_tensorpipe_set_default_timeout failing due to the
timeout being too small. Doesn't look related to this change.

Reviewers:

mrshenli

Subscribers:

Tasks:

Tags:

Fixes #67626


cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+            timeout=timeout
+        )
+
+    request_blocking = rpc_api != rpc_async
```
Even `rpc_remote` should be non-blocking, it seems like we set blocking=False below only for `rpc_async`?

kumpera replied:
Good point, let me check that.

kumpera replied:
So, while `rpc.remote()` is asynchronous, the current setup makes is particularly complex to make the RRef proxy setup fully async because the way the function target is resolved.

To make it work, we'd need to change `rpc.remote()`'s `func` argument to take either a `callable` or a `Future[Callable]`. The late would handle the fact that the type information needs to be fetched first.

pritamdamania87 replied:
> To make it work, we'd need to change rpc.remote()'s func argument to take either a callable or a Future[Callable]. The later would handle the fact that the type information needs to be fetched first.

Not sure I followed this completely, but doesn't rpc.remote()'s func argument already support a callable: https://github.com/pytorch/pytorch/blob/master/torch/distributed/rpc/api.py#L498?

kumpera replied:
It would need to take a `Future[Callable]` additionally. IE, the type would be `Callable | Future[Callable]`.
Either that or another way to delay resolving the remote type.

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+
+    # _get_type return type depends on how the first call is made to it
+    # so it's possible to get a Future even if you call it with `blocking=True`
+    is_future = issubclass(type(rref_val), torch._C.Future)
```
Isn't `is_future` basically `!request_blocking`?

kumpera replied:
Unfortunately no.

_get_type can return either `type(Future)` and `type(Class)` regardless of whether the caller requested a blocking call or not.
This behavior of `_get_type` is counter intuitive, do you think it makes sense to fix it instead of having this confusing piece of code here?

pritamdamania87 replied:
Right but there are two cases:

1. blocking=True, in this case `_get_type` returns Class (is_future=False)
2. blocking=False, in this case `_get_type` returns Future (is_future= True)

So isn't `is_future = !request_blocking`?

kumpera replied:
The problem is that `_get_type` behavior is surprising and doesn't follow what you'd rightfully assumed.

`_get_type()` can return `Class` or `Future` *regardless* of the `blocking` argument.
You can try this snippet: https://gist.github.com/kumpera/4a9560af769cd59aa2f955afb8b1de49
It prints:
```
[sync-first] blocking=True -> <class 'type'> (should be type)
[sync-first] blocking=False -> <class 'type'> (should be Future)
[async-first] blocking=False -> <class 'torch._C.Future'> (should be Future)
[async-first] blocking=True -> <class 'torch._C.Future'> (should be type)
```

Which makes sense once you see how `PyRRef::getRRefType` is implemented: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/py_rref.cpp#L271

It caches the value straight from the rpc call and doesn't try to compensate when the cached one came from a different `blocking` value than a just requested value.

One simplification would be to always request it with `blocking=False` and have the sync path call `wait`.

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+    # this function returns either the value or a Future depending on what rpc_api is.
+    def _rref_type_cont(future, is_future):
+        if is_future:
+            rref_type = future.wait()
```
Don't we need to chain callbacks here instead of just calling `wait()` on the future? Also, isn't `is_future` always going to be `False` in this code?

kumpera replied:
This callback is the one chained from the first future, we need to call `wait()`.
Calling `wait()` from the callback to extract the value is the suggested way to do so.
`is_future` is false when following the blocking path.

pritamdamania87 replied:
Ah I see, basically `future.wait()` won't wait at all. In that case, I think just for readability sake, it would be better to use `future.value()` since the invariant here is that the future has completed. Now if this invariant is not true for some reason (bugs) this would throw an error instead of blocking a thread with `future.wait()`.

kumpera replied:
Makes sense, I'll update the docs for `then` separaredly.

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+    if request_blocking or not is_future:
+        return _rref_type_cont(rref_val, is_future=is_future)
+    else:
+        return _unwrap_future(rref_val.then(lambda fut: _rref_type_cont(fut, is_future=True)))
```
Shouldn't `return rref_val.then(..)` be sufficient here? Why do we need the unwrapping?

kumpera replied:
`then` produces a Future to the return value of its callback and he callback in this case is rpc_async is produces a Future itself, so we'd end up with a `Future[Future[T]]` to the target.

pritamdamania87 replied:
Ah I see, so instead of doing this unwrapping can we do this instead:

1. Create a future `res` in this function and pass it into the function `_rref_type_cont`. 
2. Then inside `_rref_type_cont`, assign the return value of `rpc_api` to `res`.
3. Finally return `res` from this method.

kumpera replied:
Done, I inlined _unwrap_future in its call site.

rohan-varma commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+
+    # _get_type return type depends on how the first call is made to it
+    # so it's possible to get a Future even if you call it with `blocking=True`
+    is_future = issubclass(type(rref_val), torch._C.Future)
```
is `isinstance(rref_val, Future)` not a sufficient check here?

kumpera replied:
No, it's not. See the above comment on the odd behavior of `_get_type` vis-a-vis `blocking`

rohan-varma commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -12,27 +14,60 @@ def _local_invoke(rref, func_name, args, kwargs):
+        )
+
+    request_blocking = rpc_api != rpc_async
+    rref_val = rref._get_type(timeout=timeout, blocking=request_blocking)
```
It might be easier to make `_get_type` always return a future to reduce special casing here (and in other callers of this API?) If so, we can feel free to make this change in a follow up PR.

kumpera replied:
Yeah, always returning a future would eliminate most of the complexity here. I'll adjust this change to assume this will be the future behavior by always calling with `blocking=False`.

pritamdamania87 commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -13,26 +15,54 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
+        # A little explanation on this.
+        # rpc_async returns a Future pointing to the return value of `func_name`, it returns a `Future[T]`
+        # Calling _rref_type_cont from the `then` lambda causes Future wrapping. IOW, `then` returns a `Future[Future[T]]`
+        # To address that, we return a Future that is completed with the result of the async call.
+        result: Future = Future()
+
+        def _wrap_rref_type_cont(fut):
+            try:
+                _rref_type_cont(fut).then(_complete_op)
+            except BaseException as ex:
+                result.set_exception(ex)
+
+
+        def _complete_op(fut):
+            try:
+                result.set_result(fut.value())
+            except BaseException as ex:
+                result.set_exception(ex)
+
+        rref_fut.then(lambda fut: _wrap_rref_type_cont(fut))
```
I think this can be simplified as follows:

```
# Pass result to _rref_type_cont
result: Future = Future()
rref_fut.then(lambda fut: _rref_type_cont(fut, result))
return result

# Inside rref_type_cont

fut = rpc_api(
            rref.owner(),
            _invoke_func,
            args=(rref, func_name, args, kwargs),
            timeout=timeout
        )
 fut.then(lambda fut: result.set_result(fut.value()))
```

kumpera replied:
That approach won't handle any exceptions thrown in `_rref_type_cont`'s body nor would it propagate failures.

AFAICT, this is how you copy the outcome of one future to another:
```
def copy_future_outcome(from, to):
try:
   to.set_result(from.value())
except BaseException as e:
   from.set_exception(e)
```

Plus rref_type_cont handles both the synchronous and asynchronous path, so we'd need to check for that before consuming the result of the `rpc_api` call.

pritamdamania87 replied:
One way to deal with that is to have a try-except clause in _rref_type_cont body itself and then set the exception on `result`, similary for `result.set_result(fut.value())`. Will leave this up to you though.

rohan-varma commented on torch/distributed/rpc/rref_proxy.py:
```diff
@@ -13,26 +15,54 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
+            timeout=timeout
+        )
+
+    request_blocking = rpc_api != rpc_async
```
Looks like it is unused now?

kumpera replied:
yes, it's dead. Just pushed a fix for that.

rohan-varma reviewed (approved):
Looks good to me, although this code is also used internally, would be good to double check any callsites (everything should be covered by tests though anyways).

kumpera committed:
[distributed] Make rref_proxy._invoke_rpc trully async when needed. (#70206)

Summary:
From https://github.com/pytorch/pytorch/issues/67626: RRefProxy (rref.rpc_async, rref.rpc_sync, rref.remote) currently uses a blocking RPC call to the owner

This is done by chaining async calls. In the sync case we wait on the
resulting Future.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/70206

Test Plan:
I ran rpc_tests using tensorpipe_rpc_agent_test_fixture.py and had to
adjust test_rref_proxy_timeout to the new behavior.

I ran into test_tensorpipe_set_default_timeout failing due to the
timeout being too small. Doesn't look related to this change.
mrshenli
Fixes https://github.com/pytorch/pytorch/issues/67626

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse SciPioneer H-Huang

Reviewed By: pritamdamania87

Differential Revision: D33243348

Pulled By: kumpera

fbshipit-source-id: cd2d2c27ef966dd646b87bfd6f50ffcb2f6b9010
